snippets/core/system-text-json/csharp/RoundtripToFileAsync.cs

```diff
@@ -29,6 +30,24 @@ public static async Task RunAsync()
             }
             // </SnippetDeserialize>
             weatherForecast.DisplayPropertyValues();
+
+            using (FileStream fs = File.Create(fileNameUtf8))
+            {
+                using (StreamWriter writer = new StreamWriter(fs, Encoding.UTF8))
+                {
+                    await JsonSerializer.SerializeAsync(fs, weatherForecast);
+                }
+            }
+            Console.WriteLine($"The result is in {fileNameUtf8}\n");
+            using (FileStream fs = File.OpenRead(fileNameUtf8))
+            {
+                //System.Text.Json.JsonException: '0xEF' is invalid after a single JSON value. Expected end of data.
+                //weatherForecast = await JsonSerializer.DeserializeAsync<WeatherForecast>(fs);
```

Comment:
Why is this commented out? Are you getting a `JsonException`?

Reply:
Yes, I was getting an exception when trying to read a file encoded as UTF-8 with the preceding code.

Reply:
Are you sure? That would be a bug. I just tried it and the UTF-8 BOM bytes are correctly getting ignored. Can you share the repro file and sample where you are seeing this issue?

Reply:
These samples are in a project that runs through all of the samples when you run it. If you have this branch locally you can open the project and uncomment the `weatherForeast=await...` line, run the project, and you'll see the error.

Reply:
I'm pushing a commit that will eliminate this code that throws the exception - if you haven't tried it and still want to, get the code at this commit -- https://github.com/tdykstra/dotnet-samples/tree/b1c945b488b5966f816e92ddac7c002f9656b635/snippets/core/system-text-json/csharp

Reply:
When I try to run it, I get the following errors:
```text
E:\GitHub\Fork\samples\snippets\core\system-text-json\csharp>dotnet run

RoundtripDictionaryTkeyEnumTValue.cs(6,18): error CS0101: The namespace 'SystemTextJsonSamples' already contains a definition for 'RoundtripDictionaryTkeyEnumTValue' [E:\GitHub\Fork\samples\snippets\core\system-text-json\csharp\SystemTextJsonSamples.csproj]
DeserializeInferredTypesToObject.cs(6,18): error CS0101: The namespace 'SystemTextJsonSamples' already contains a definition for 'DeserializeInferredTypesToObject' [E:\GitHub\Fork\samples\snippets\core\system-text-json\csharp\SystemTextJsonSamples.csproj]
SerializePolymorphic.cs(6,11): error CS0101: The namespace 'SystemTextJsonSamples' already contains a definition for 'SerializePolymorphic' [E:\GitHub\Fork\samples\snippets\core\system-text-json\csharp\SystemTextJsonSamples.csproj]
RoundtripDictionaryTkeyEnumTValue.cs(8,28): error CS0111: Type 'RoundtripDictionaryTkeyEnumTValue' already defines a member called 'Run' with the same parameter types [E:\GitHub\Fork\samples\snippets\core\system-text-json\csharp\SystemTextJsonSamples.csproj]
DeserializeInferredTypesToObject.cs(8,28): error CS0111: Type 'DeserializeInferredTypesToObject' already defines a member called 'Run' with the same parameter types [E:\GitHub\Fork\samples\snippets\core\system-text-json\csharp\SystemTextJsonSamples.csproj]
SerializePolymorphic.cs(8,28): error CS0111: Type 'SerializePolymorphic' already defines a member called 'Run' with the same parameter types [E:\GitHub\Fork\samples\snippets\core\system-text-json\csharp\SystemTextJsonSamples.csproj]
```

Reply:
samples master has a temporary change in it that was needed to fix a docs build bug but broke the project, this PR fixes it and the docs build.  Sorry for the confusion.  I should have mentioned that  you needed specifically the PR branch, which is in my fork.

Reply:
OK, I found out the issue. When you use `Encoding.UTF8` within a `StreamWriter`, it says that it is going to emit a UTF-8 BOM.
https://source.dot.net/#System.Private.CoreLib/StreamWriter.cs,81
https://source.dot.net/#System.Private.CoreLib/UTF8Encoding.cs,62

This wouldn't be a problem if the UTF-8 BOM was written at the beginning. The `JsonSerializer` ignores it, correctly.

However, because you are **not** using the `StreamWriter` and instead writing to the `FileStream` directly, the JSON payload gets written **first**, and the BOM bytes get written after, when the `StreamWriter` is disposed. The `JsonSerializer.Deserializer` then correctly throws `JsonException` since you have invalid JSON now.

Can't have byte order mark at the end of the payload (hence invalid).
![image](https://user-images.githubusercontent.com/6527137/71494158-c8f41200-27f9-11ea-9481-93855a4a767c.png)

Also note, you likely want the filename to contain the `.json` extension.
https://github.com/dotnet/samples/blob/bc25d342d4cf041b953665cb5235e8d5489d34a0/snippets/core/system-text-json/csharp/RoundtripToFileAsync.cs#L14

Removing the `StreamWriter` should and does fix the underlying issue :)

Reply:
And in case you missed the note I sent just as I received yours, the code that generates the exception is gone from the PR branch but can be found here: https://github.com/tdykstra/dotnet-samples/tree/b1c945b488b5966f816e92ddac7c002f9656b635/snippets/core/system-text-json/csharp
But given that there's no need to wrap a StreamWriter the way I was doing it, it may not be worth investigating further.

Reply:
Emails crossed again, got the resolution notice, thanks !

Reply:
Yep, sounds good. I understood what you were getting at. Looks like there is no underlying product issue and the sample has been fixed.